### DISCUSSION TITLE
Select framework for Java-based API Service implementations

### DISCUSSION THREAD

**Discussion by jeffreyscarpenter** [stargate/stargate#1526] [RFCs]
In the first milestone of Stargate v2, we separated out the REST API Service from the Stargate coordinator node as a separately deployable microservice. In the first stage of this refactoring, the service was not changed from its current configuration as a Dropwizard-based service running against Java 8.

One of the next steps in the maturation of this REST API Service is to migrate it to a more modern Java framework, which should go along with upgrading our Java version. Various frameworks have been proposed, including [Spring Boot](https://spring.io/projects/spring-boot), [Micronaut](https://micronaut.io/) and [Quarkus](https://quarkus.io/). 

This discussion is intended to kick off the effort to select a new framework to be used for the REST API service and other Java-based API Services. 

Here are a few proposed requirements or ways in which we might compare frameworks:

- Minimize startup time
- Minimize memory usage
- Minimize Docker image size
- Support for target Java version (what should our target Java version be)
- Support for gRPC/Netty (services are intended to use gRPC to communicate with Coordinator Nodes)
- Community maturity: level of adoption, support, release frequency, security patching

Are there other requirements or frameworks we should consider?

**Comment by ivansenic** [comment]
I would like to contribute to this discussion and present my personal opinions on what should be considered when selecting a new framework. Please consider the table below as the non-final, as I can not cover everything and might be adding things that pop-up in addition. Not everything is mandatory, and some items weight more than others of course. I tried to fill in the data about support in frameworks that I have from my experience and from reading the official documentation.

*Pending analysis*
* [x] Resilience & Fault tolerance

| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; | Micronaut  | SpringBoot | Quarkus  | Dropwizard (current) |
|---|---|---|---|---|
| **Build** |   |   |   |   |
| Support for Java 17  | :heavy_check_mark:  | :heavy_check_mark:  | :warning: [In progress](https://github.com/quarkusio/quarkus/issues/21481)  | :no_entry: Java 11 |
| Support for latest Maven/Gradle wrappers  | :heavy_check_mark:  | :heavy_check_mark:  | :heavy_check_mark:   | :warning: Maven only |
| Dependency management   | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:   | :no_entry: |
| **Images**  |   |   |   | |
| Docker image building | :heavy_check_mark:   |  :heavy_check_mark:  | :heavy_check_mark:   | :no_entry: |
| Graal native image building | :heavy_check_mark: :information_source:  AOT compilaton available only with the Gradle.  | :heavy_check_mark: :warning: Not sure how it all plays with Spring's reflection-based approach. Experimental AOT plugin.  | :heavy_check_mark: :information_source: Seems that reflection cases have to be manually taken care of in a configruation. | :no_entry: |
| **Dependency injection**  |   |   |   |  |
| Conditional DI | :heavy_check_mark:  | :heavy_check_mark:  | :heavy_check_mark:   | :no_entry:  Manual DI |
| Auto-configuration  | :heavy_check_mark: :question: Not sure about the extent of the available features.  | :heavy_check_mark:  | :heavy_check_mark: :warning: Class-path related conditions seems not to be possible.  | :no_entry: |
| Beans validation | :heavy_check_mark:  | :heavy_check_mark:  | :heavy_check_mark:  | :heavy_check_mark: |
| **Configuration** |   |   |   | |
| Hierarchical config (files, jvm agrs, env vars) | :heavy_check_mark:  | :heavy_check_mark:  |  :heavy_check_mark:  | :heavy_check_mark: |
| Support for profiles/environments | :heavy_check_mark:  | :heavy_check_mark:  |  :heavy_check_mark:  | :no_entry: |
| Configuration model validation | :heavy_check_mark:  | :heavy_check_mark:  |  :heavy_check_mark:  | :heavy_check_mark: |
| **Integrations & Extensions** |   |   |   | |
| Reactive support |  Reactor, RxJava | Reactor  | Mutiny  | :no_entry: |
| Reactive HTTP CS | :heavy_check_mark:  | :heavy_check_mark:  | :heavy_check_mark:   | :no_entry: Async servlet |
| Reactive gRPC CS | :heavy_check_mark:   |  :no_entry: SpringBoot has no official gRPC support. External contributions exist.  |  :heavy_check_mark:  | :no_entry: |
| gRPC support |  :heavy_check_mark:   |  :no_entry: SpringBoot has no official gRPC support. External contributions exist.  |  :heavy_check_mark:  | :no_entry: |
| GraphQL support |  :heavy_check_mark:   |  :warning: In a milestone phase, by GraphQL Java team.  |  :heavy_check_mark:  | :no_entry: |
| JAX-RS support | :heavy_check_mark:  | :heavy_check_mark:  | :heavy_check_mark:  | :heavy_check_mark: |
| **Observability** |   |   |   | |
| Management endpoints (health, etc) | :heavy_check_mark:  | :heavy_check_mark:  | :heavy_check_mark:   | :heavy_check_mark: |
| Micrometer support with Prometheus registry | :heavy_check_mark:  | :heavy_check_mark:  | :heavy_check_mark:   | :warning: No micrometer support |
| Metrics (JVM, HTTP, gRPC)  | :heavy_check_mark: :warning: No integration for gRPC metrics.  | :heavy_check_mark: :warning: No integration for gRPC metrics.  | :heavy_check_mark:   | :heavy_check_mark: :warning: No integration for gRPC metrics. |
| Tracing (HTTP, gRPC)  | :heavy_check_mark:   | :heavy_check_mark: :warning: No integration for gRPC tracing.  | :heavy_check_mark:   | :no_entry:  |
| **Security** |   |   |   |  |
| Header based authentication | :warning:   Must implement custom, but fairly easy.  | :heavy_check_mark:   | :warning: Custom programming possible.  | :warning:   Must implement custom, but fairly easy.|
| **Testing** |   |   |   |  |
| Integration tests (WireMock, gRPC, etc) | :heavy_check_mark: :warning: No WireMock integration.  | :heavy_check_mark: :warning: No  gRPC integration.  | :heavy_check_mark: | :heavy_check_mark: :warning: No WireMock, no gRPC integration. |
| **Cloud native features** |   |   |   |
| Kubernetes support | :heavy_check_mark:   | :heavy_check_mark:   | :heavy_check_mark:   | :no_entry: |
| **Other** |   |   |   |  |
| Open API 3 integration (Swagger) | :heavy_check_mark:   | :heavy_check_mark:   | :heavy_check_mark:   | :warning: Manual |
| Open API 3 Stub Generator | :no_entry: [3rd party only](https://github.com/kokuwaio/micronaut-openapi-codegen)  | :heavy_check_mark:  | :heavy_check_mark: :warning: Vert.x only  | :heavy_check_mark: |
| Cache abstraction | :heavy_check_mark:   | :heavy_check_mark:   |  :heavy_check_mark:  | :no_entry: |
| Scheduling support | :heavy_check_mark:   | :heavy_check_mark:   | :heavy_check_mark:  | :no_entry: |
| Dev tools | :heavy_check_mark: CLI, IntelliJ, reloading  | :heavy_check_mark: CLI, IntelliJ, `spring-boot-devtools`  | :heavy_check_mark: CLI, IntelliJ, reloading  | :no_entry: No dedicated dev tool. |
| **General** |   |   |   |  |
| Extensive documentation | :warning: Documentation often feels incomplete.  | :heavy_check_mark:   |  :heavy_check_mark:  Spread in guides, full reference as PDF.  | :heavy_check_mark: Existing features documented. |
| Community maturity | :question:   | :heavy_check_mark:   | :question:   | :question:  |
| Stargate contributors experience | :no_entry: @ivansenic <br/>:question: @mpenick<br/>:question: @olim7t<br/>:question: @jeffreyscarpenter<br/>:question: @tatu-at-datastax<br/>:question: @tomekl007<br/>:question: @EricBorczuk    |  :heavy_check_mark: @ivansenic <br/>:question: @mpenick<br/>:question: @olim7t<br/>:question: @jeffreyscarpenter<br/>:heavy_check_mark: @tatu-at-datastax<br/>:heavy_check_mark: @tomekl007<br/>:question:  @EricBorczuk | :no_entry: @ivansenic <br/>:question: @mpenick<br/>:question: @olim7t<br/>:question: @jeffreyscarpenter<br/>:question: @tatu-at-datastax<br/>:heavy_check_mark: @tomekl007<br/>:question:  @EricBorczuk | :heavy_check_mark:  |

### Recap

#### Micronaut
Seems that biggest pros of the Micronaut is in the support for the gRPC, not only with the wrappers, but with tracing and testing as well. The native image support with the AOT compilation should provide very good if not the best startup times and memory consumption. 

It feels that security support is limited, but then again it's quite questionable if we would be able to use any support here at all. Documentation feels personally to me often too short and limited, but is definitely good for general usage and covers basic use-cases.

#### SpringBoot
SpringBoot should have best documentation, examples and community around. I am pretty sure the extensive security features it offers would win over other frameworks, but it's questionable how much of that we can use. It lacks official support for the gRPC and since our communication protocol will be gRPC this seems like a problem here. 

Reading some of the framework comparisons pointed that Spring has worst startup times and highest memory consumption.

It's expected that the developers in general have the most experience with SpringBoot, due to it's huge adoption. It's also more likely that Spring would have support for any features we might need in the future and that were not recognized in the list above.

#### Quarkus
Quarkus has a complete support for gRPC, including the full observability. Seems that all major features are available, except the auto-configuration based on the class path.

I feel that the biggest con for Quarkus is that it's based on the Multiny reactive framework. The other two frameworks are taking Project Reactor as the main framework and it seems that Reactor is becoming the de-facto standard with a lot of great features. It's unknown how experienced are Stargate contributors with the Multiny framework and what is learning curve gonna be.

It's unclear what's the current state with the Java 17 support and when it will be fully supported. It's also unclear how it compares against other frameworks wrt the native image building. 


## Documentation / Further Reading

### Micronaut

([Docs](https://micronaut.io/docs/))

* [Reactive REST](https://docs.micronaut.io/latest/guide/#reactiveServer) 
* [gRPC client](https://micronaut-projects.github.io/micronaut-grpc/latest/guide/)
* [GraphQL support](https://micronaut-projects.github.io/micronaut-graphql/latest/guide/) 
* [Image building](https://docs.micronaut.io/latest/guide/#deployingApp)
* [Metrics Prometheus](https://micronaut-projects.github.io/micronaut-graphql/latest/guide/)
* [Tracing](https://docs.micronaut.io/latest/guide/#distributedTracing)
* [Fault tolerance](https://docs.micronaut.io/latest/guide/#retry)
* [OpenAPI/Swagger](https://micronaut-projects.github.io/micronaut-openapi/latest/guide/index.html)

### Quarkus

([All Gudes](https://quarkus.io/guides/))

* [Reactive REST](https://quarkus.io/guides/resteasy-reactive) 
* [gRPC client](https://quarkus.io/guides/grpc-service-consumption)
* [GraphQL support](https://quarkus.io/guides/smallrye-graphql) 
* [Image building](https://quarkus.io/guides/container-image)
* [Metrics Prometheus](https://quarkus.io/guides/micrometer)
* [Tracing](https://quarkus.io/guides/opentracing)
* [Fault tolerance](https://quarkus.io/guides/smallrye-fault-tolerance)
* [OpenAPI/Swagger](https://quarkus.io/guides/openapi-swaggerui)

**Reply by polandll** [reply]
I see that all 3 options support Open API 3 integration (Swagger), but I would like to add Swagger annotations as a selection criteria as well, if possible. We (Docs) would like to add extensive swagger annotations to the source code, so that we can generate proper OS yaml files directly from the source code. Currently, we have to heavily edit the yaml file that is generated by swagger.

**Reply by polandll** [reply]
Alternatively, the source code could use an oas.yaml file directly. I believe cloud-ondemand does that: 
https://github.com/riptano/cloud-ondemand/tree/74da78c1f9363fb5fa77042e26cd08ecc60f60b6/orchestrator-client/pkg/client/v2/api

**Reply by ivansenic** [reply]
@polandll Wait are those annotations available in the Open API integrations? From my experience those are like:

```
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.enums.ParameterIn;
import io.swagger.v3.oas.annotations.headers.Header;
import io.swagger.v3.oas.annotations.media.ArraySchema;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
```
At least from my previous SpringBoot projects those are the ones that you get with:

```
// openapi
implementation 'org.springdoc:springdoc-openapi-core:1.+'
implementation 'org.springdoc:springdoc-openapi-webmvc-core:1.+'
implementation 'org.springdoc:springdoc-openapi-data-rest:1.+'
```

We can further check this, but I guess every framework supports this.. For example, [Micronaut](https://micronaut-projects.github.io/micronaut-openapi/latest/guide/index.html) points to the same, as well as [Quarkus](https://quarkus.io/guides/openapi-swaggerui) apparently.

**Comment by tatu-at-datastax** [comment]
Thank you @ivansenic for the comprehensive list of criteria. That should help in figuring out what choice to make evaluating the best choice as the future framework. All the notes make sense and align with what I have heard of frameworks (my experience is limited to some use of Spring Boot; I have only read about the other 2 newer ones).

One thing that might be worth noting is that the current framework (DropWizard/Jersey/JAX-RS) does not have in-built support for gRPC or Reactive processing (I think), but Stargate uses both. So in a way explicit framework support is not mandatory for either, I suspect, although can be a big plus.

**Reply by ivansenic** [reply]
@tatu-at-datastax I would actually argue that explicit framework support is somehow mandatory.. Or at least expected..

The first reason is the dependency management. If you don't have to bring a 3rd party library to support something, that you can really on the framework releases to cover needed updates for that specific feature(s). Meaning you don't have to maintain compatibility of the 3rd party solution and update it yourself. Furthermore, integrated solution exposed by a framework will also be way more tested as it's an official solution, then if you bring extra library.

Second thing is the free lunch you get wrt other concerns. Take gRPC for example, bring your own 3rd party solution means you need now to integrate also metrics and tracing for that library into the existing concepts of the framework.. That's extra work that you need to invest in, instead of relying on again well tested integrated options that framework expose.

**Reply by tatu-at-datastax** [reply]
All I am saying is that we have managed to get things working without in-built support. Not that we should ignore integration opportunities, or that there weren't any benefits.

**Comment by tatu-at-datastax** [comment]
Should we consider JAX-RS (Annotation) support as well? Spring Boot has it, trying to figure out if Micronaut has it explicitly (or requires use of annotation mapper)

**Reply by ivansenic** [reply]
What would `JAX-RS support` mean? Support for the annotations?

**Reply by tatu-at-datastax** [reply]
Correct: support use of JAX-RS annotations (instead of framework native ones). I don't think we rely too heavily on other JAX-RS functionality.

**Reply by ivansenic** [reply]
But why would this be important?

**Reply by tatu-at-datastax** [reply]
Our current API is defined using those annotations, and they are sort of standard in Java world. So it seems beneficial to be able to use those. Not a blocker but if using framework-specific annotations there is more work to convert to the framework (and if we didn't like it, away).
So just for interoperability, easier for new contributors to learn how things work.

Micronaut has some level of support for translating JAX-RS annotations, f.ex:

https://github.com/micronaut-projects/micronaut-jaxrs

**Reply by ivansenic** [reply]
I see what s your point, you don't want to remap the code to the framework-specific annotations.. And I get you.. But I guess this is not such a hard work, and some changes will be needed for sure.. But I can add it to the list of the features :+1:

**Reply by tatu-at-datastax** [reply]
Sounds good to me! I definitely would not consider this a must-have.

**Comment by tatu-at-datastax** [comment]
One suggestion: since most of team members are probably not very familiar with Quarkus or Micronaut, we could add links to documents/articles that give good overview of their usage. This could help everyone get familiar with kinds of changes, improvements, beyond just listing functionality offered. At least for me watching presentations about Micronaut has been helpful in realizing how big the changes are, wrt reducing work needed (but also changing to use more and more indirection via annotations, naming conventions).
I'll see if I can add some.

Fine to add Spring Boot ones too of course.

**Reply by ivansenic** [reply]
Sure why not... Here are some of the interesting docs from my point of view:
 
* Micronaut ([Docs](https://micronaut.io/docs/))
   * [Reactive REST](https://docs.micronaut.io/latest/guide/#reactiveServer) 
   * [gRPC client](https://micronaut-projects.github.io/micronaut-grpc/latest/guide/)
   * [GraphQL support](https://micronaut-projects.github.io/micronaut-graphql/latest/guide/) 
   * [Image building](https://docs.micronaut.io/latest/guide/#deployingApp)
   * [Metrics Prometheus](https://micronaut-projects.github.io/micronaut-graphql/latest/guide/)
   * [Tracing](https://docs.micronaut.io/latest/guide/#distributedTracing)
   * [Fault tolerance](https://docs.micronaut.io/latest/guide/#retry)
   * [OpenAPI/Swagger](https://micronaut-projects.github.io/micronaut-openapi/latest/guide/index.html)
* Quarkus ([All Gudes](https://quarkus.io/guides/))
   * [Reactive REST](https://quarkus.io/guides/resteasy-reactive) 
   * [gRPC client](https://quarkus.io/guides/grpc-service-consumption)
   * [GraphQL support](https://quarkus.io/guides/smallrye-graphql) 
   * [Image building](https://quarkus.io/guides/container-image)
   * [Metrics Prometheus](https://quarkus.io/guides/micrometer)
   * [Tracing](https://quarkus.io/guides/opentracing)
   * [Fault tolerance](https://quarkus.io/guides/smallrye-fault-tolerance)
   * [OpenAPI/Swagger](https://quarkus.io/guides/openapi-swaggerui)

@olim7t @tomekl007 Totally forgot to add the GraphQL to the table, please check the guides above and see if this brings any benefit to you.

**Comment by jdonenine** [comment]
One section of the table above that's empty at the moment is observability -- I don't have any experience with Micronaut or Qurkus on that front but I've found what Boot provides through actuator and the potential to easily bolt in the spring-boot-admin application really quite complete.  It should provide what's necessary for K8s probes out of the box as well:  https://docs.spring.io/spring-boot/docs/current/reference/html/actuator.html#actuator.endpoints.kubernetes-probes

**Reply by ivansenic** [reply]
Observability is the section with 4 items below (metrics, tracing, etc).. I think I made confusion with bold text, will fix it.. Bold text -> section header..

**Comment by jdonenine** [comment]
Also thanks for putting up this convo @jeffreyscarpenter @tatu-at-datastax @ivansenic I think this type of discussion can be really useful for a lot of different folks who are facing decisions like this in many different projects, not just ours here.
